File: web/src/pages/spark/sparkDuration.ts
/**
 * The Spark countdown duration domain.
 *
 * The countdown used to be the literal `60` in SparkTimer plus the phrase "one
 * minute" repeated through card copy, button labels, the static library and the
 * system prompt. Now that a participant sets it, exactly one thing may name a
 * duration -- the resolved value -- and everything else asks {@link formatDuration}.
 *
 * Mirrors `api/app/services/spark_duration.py`. The bounds live in both places
 * because the server cannot trust the browser and the browser cannot wait for
 * the server to reject a chip tap; the wire policy is what reconciles them.
 */

declare const durationBrand: unique symbol;

/** Whole seconds within the study bounds. Construct via {@link parseDuration}. */
export type SparkDuration = number & { readonly [durationBrand]: true };

/**
 * Who chose the duration. Closed, and deliberately without a "the model chose
 * it" variant: nothing populates one today, and a dead variant is a state the
 * analysis has to defend against for no benefit.
 */
export type DurationSource = "study_default" | "participant";

export interface ResolvedDuration {
    readonly seconds: SparkDuration;
    readonly source: DurationSource;
}

export const MIN_DURATION_SECONDS = 15;
export const MAX_DURATION_SECONDS = 300;

/** Used only until the first generate response carries the real policy. */
export const FALLBACK_DEFAULT_SECONDS = 60;
const FALLBACK_CHOICES = [30, 60, 90, 120, 180, 300];

/** What the participant may pick from, as served by the API. */
export interface TimerPolicy {
    readonly defaultSeconds: SparkDuration;
    readonly choices: readonly SparkDuration[];
}

/**
 * Admit an untrusted value, or `null` when it is not a duration.
 *
 * The one place a duration brand is asserted. Callers never cast.
 */
export function parseDuration(raw: unknown): SparkDuration | null {
    if (typeof raw !== "number" || !Number.isInteger(raw)) return null;
    if (raw < MIN_DURATION_SECONDS || raw > MAX_DURATION_SECONDS) return null;
    return raw as SparkDuration;
}

/** Parse a duration, falling back to the study default when it is unusable. */
export function parseDurationOr(raw: unknown, fallback: SparkDuration): SparkDuration {
    return parseDuration(raw) ?? fallback;
}

// Safe by construction: FALLBACK_DEFAULT_SECONDS is a literal inside the bounds
// declared directly above it, so parseDuration cannot reject it. The assertion
// removes the null, not the domain check.
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const FALLBACK_DEFAULT = parseDuration(FALLBACK_DEFAULT_SECONDS)!;

/**
 * Parse the timer policy off a generate response.
 *
 * Total: a missing, malformed, or entirely out-of-range policy yields the
 * built-in fallback rather than an exception, because a countdown the
 * participant cannot start is a worse failure than a countdown of the wrong
 * length. Out-of-domain individual choices are dropped, not clamped -- clamping
 * would silently offer a length the study never sanctioned.
 */
export function parseTimerPolicy(raw: unknown): TimerPolicy {
    const source = (raw ?? {}) as { default_seconds?: unknown; choices?: unknown };
    const defaultSeconds = parseDurationOr(source.default_seconds, FALLBACK_DEFAULT);
    const rawChoices = Array.isArray(source.choices) ? source.choices : FALLBACK_CHOICES;
    const choices = rawChoices
        .map(parseDuration)
        .filter((d): d is SparkDuration => d !== null);
    return {
        defaultSeconds,
        // A policy whose every choice was rejected still has to offer the
        // default, or the control renders empty and nothing can be started.
        choices: choices.length > 0 ? choices : [defaultSeconds],
    };
}

/** Total: an explicit participant pick wins, otherwise the study default. */
export function resolveDuration(
    studyDefault: SparkDuration,
    participantChoice: SparkDuration | null,
): ResolvedDuration {
    return participantChoice === null
        ? { seconds: studyDefault, source: "study_default" }
        : { seconds: participantChoice, source: "participant" };
}

/**
 * Human-readable length. Whole minutes read as minutes, everything else as
 * seconds -- "90 seconds" is clearer than "1.5 minutes".
 */
export function formatDuration(seconds: SparkDuration): string {
    if (seconds % 60 !== 0) return `${seconds} seconds`;
    const minutes = seconds / 60;
    return minutes === 1 ? "1 minute" : `${minutes} minutes`;
}

File: api/app/services/spark_duration.py
"""The Spark countdown duration domain.

The countdown used to be the literal ``60`` in ``SparkTimer.tsx`` plus the
phrase "one minute" scattered through card copy, the static library and the
system prompt. Making it participant-configurable means exactly one of those
may name a duration -- the resolved value -- so this module owns the number and
every consumer asks it.

Two things can choose a duration, and they are a closed set:

``study_default``
    The researcher-configured constant every flow starts from.
``participant``
    An explicit pick from :data:`DURATION_CHOICES`.

There is deliberately no "the model chose it" variant. Nothing populates one
today, and a dead variant is a state the analysis has to defend against for no
benefit. Adding it later is a compiler-guided edit: every ``match`` over
:data:`DurationSource` fails until it is handled.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Self

# Absolute bounds on any duration the domain will admit. Below 15s the ring
# animation and the "move until it ends" framing stop meaning anything; above
# 300s it is no longer a micro-break, which is the whole intervention.
MIN_DURATION_SECONDS = 15
MAX_DURATION_SECONDS = 300

#: The study default, and the value every pre-configurable-duration research
#: row should be read as carrying.
DEFAULT_DURATION_SECONDS = 60

#: What a participant may pick from. A presentation policy, deliberately
#: narrower than the domain bounds: the type admits any valid second count, the
#: UI offers a short row of chips.
DURATION_CHOICES: tuple[int, ...] = (30, 60, 90, 120, 180, 300)

DurationSource = Literal["study_default", "participant"]


@dataclass(frozen=True, slots=True)
class SparkDuration:
    """A countdown length in whole seconds, known to be within bounds.

    Construct through :meth:`parse`. ``__post_init__`` re-checks because Python
    cannot hide the constructor, so direct construction raises rather than
    admitting an out-of-range value to the domain.
    """

    seconds: int

    def __post_init__(self) -> None:
        if not _in_domain(self.seconds):
            raise ValueError(
                f"Spark duration must be a whole number of seconds in "
                f"[{MIN_DURATION_SECONDS}, {MAX_DURATION_SECONDS}]: {self.seconds!r}"
            )

    @classmethod
    def parse(cls, raw: object) -> Self | None:
        """Admit an untrusted value, or ``None`` when it is not a duration.

        ``bool`` is rejected explicitly: it is an ``int`` subclass in Python, so
        ``SparkDuration.parse(True)`` would otherwise be a 1-second countdown.
        """
        if isinstance(raw, bool) or not isinstance(raw, int):
            return None
        return cls(raw) if _in_domain(raw) else None


def _in_domain(seconds: int) -> bool:
    return MIN_DURATION_SECONDS <= seconds <= MAX_DURATION_SECONDS


@dataclass(frozen=True, slots=True)
class ResolvedDuration:
    """A duration paired with who chose it.

    Both halves travel together because neither is interpretable alone: a
    research row recording 180 seconds without recording that the participant
    asked for it cannot be separated from a study whose default was 180.
    """

    duration: SparkDuration
    source: DurationSource


def resolve_duration(
    study_default: SparkDuration,
    participant_choice: SparkDuration | None,
) -> ResolvedDuration:
    """Total: participant choice wins, otherwise the study default."""
    if participant_choice is None:
        return ResolvedDuration(study_default, "study_default")
    return ResolvedDuration(participant_choice, "participant")


def study_default_duration() -> SparkDuration:
    """The configured study default as a parsed domain value."""
    return SparkDuration(DEFAULT_DURATION_SECONDS)
